src/bridges/erc4626/VaultBridge.sol

```diff
@@ -0,0 +1,157 @@
+     * executed. If input is the vault share token, unstaking. RollupProcessor.sol has to transfer the tokens to the bridge before calling
+     * the method. If this is not the case, the function will revert (either in STAKING_CONTRACT.stake(...) or during
+     * SB burn).
+     *
+     * @param inputAssetA - Vault asset (Staking) or Vault Address (Unstaking)
+     * @param outputAssetA - Vault Address (Staking) or Vault asset (UnStaking)
+     * @param totalInputValue - the amount of assets to stake or shares to unstake
+     * @return outputValueA - the amount of shares (Staking) or assets (Unstaking) minted/transferred to
+     * the RollupProcessor.sol
+     */
+    function convert(
+        AztecTypes.AztecAsset memory inputAssetA,
+        AztecTypes.AztecAsset memory inputAssetB,
+        AztecTypes.AztecAsset memory outputAssetA,
+        AztecTypes.AztecAsset memory outputAssetB,
+        uint256 totalInputValue,
+        uint256 interactionNonce,
+        uint64 auxData,
+        address rollupBeneficiary
+    )
+        external
+        payable
+        override
+        returns (
+            uint256 outputValueA,
+            uint256 outputValueB,
+            bool isAsync
+        )
+    {
+        // // ### INITIALIZATION AND SANITY CHECKS
+        require(msg.sender == rollupProcessor, 'ExampleBridge: INVALID_CALLER');
+
+        if (testPair(outputAssetA.erc20Address, inputAssetA.erc20Address)) {
+            outputValueA = enter(outputAssetA.erc20Address, inputAssetA.erc20Address, totalInputValue);
+        } else if (testPair(inputAssetA.erc20Address, outputAssetA.erc20Address)) {
+            outputValueA = exit(inputAssetA.erc20Address, outputAssetA.erc20Address, totalInputValue);
+        } else {
+            revert('invalid vault token pair address');
+        }
+    }
+
+    /**
+     * @notice Function which test whether a token is a vaild asset for a given vault
+     * @dev This method is used to check whether input and output assets are matching vault share/asset pairs
+     *
+     * @param a address of erc4626 vault
+     * @param b address of the vault asset token
+     *
+     */
+    function testPair(address a, address b) public returns (bool) {
+        try IERC4626(a).asset() returns (IERC20 token) {
+            return address(token) == b;
+        } catch {
+            return false;
+        }
+    }
+
+    /**
+     * @notice Internal Function used to stake
+     * @dev This method deposits an exact amount of asset into an erc4626 vault then approves the rollup processor
+     *
+     * @param vault address of erc4626 vault
+     * @param token address of the vault asset token
+     * @param amount amount of an asset to be burned
+   
+     */
+    // enter by deposit exact assets
+    function enter(
+        address vault,
+        address token,
+        uint256 amount
+    ) internal returns (uint256) {
+        IERC20 depositToken = IERC20(token);
+        depositToken.approve(vault, amount);
```

Comment:
It would be useful to pre-approve these tokens to safe gas. The bridge should never hold any ERC20s after the interaction settles so it is not a security risk. Please check our [liquity bridge](https://github.com/AztecProtocol/aztec-connect-bridges/blob/e53c3f2fe746aa02f680571f29c08f02a90d932d/src/bridges/liquity/StakingBridge.sol#L70) for inspiration.

Reply:
That makes sense. The bridge I wrote i designed to work for any vault/ token pair.It is obviously impossible to pre approve them all. I could make a function that allows pre approvals for specific pairs. Would that make sense?